erts/emulator/beam/erl_gc.c

```diff
@@ -3821,6 +3823,165 @@ erts_dbg_within_proc(Eterm *ptr, Process *p, Eterm *real_htop)
+                if (is_magic_ref_thing(rtp)) {
+                    goto off_heap_common;
+                }
+                /* Fall through... */
+            }
+            default:
+                {
+                    int tari = header_arity(val);
+                    tp += tari;
+                }
+                break;
+            }
+            break;
+        }
+    }
+    return 1;
+}
+
+int
+erts_dbg_check_heap_terms(int (*check_eterm)(Eterm),
+                          Process *p,
+                          Eterm *real_htop)
+{
+    ErlHeapFragment* bp;
+    ErtsMessage* mp;
+    Eterm *htop, *heap;
+
+    if (p->abandoned_heap) {
+	ERTS_GET_ORIG_HEAP(p, heap, htop);
+	if (!check_all_heap_terms_in_range(check_eterm,
+                                           heap, htop))
+	    return 0;
+    }
+
+    heap = p->heap;
+    htop = real_htop ? real_htop : HEAP_TOP(p);
+
+    if (OLD_HEAP(p) &&
+        !check_all_heap_terms_in_range(check_eterm,
+                                       OLD_HEAP(p), OLD_HTOP(p) /*OLD_HEND(p)*/)) {
+        return 0;
+    }
+
+    if (!check_all_heap_terms_in_range(check_eterm, heap, htop)) {
+        return 0;
+    }
+
+    mp = p->msg_frag;
+    bp = p->mbuf;
+
+    if (bp)
+	goto search_heap_frags;
+
+    while (mp) {
+
+        bp = erts_message_to_heap_frag(mp);
+	mp = mp->next;
+
+    search_heap_frags:
+
+	while (bp) {
+            if (!check_all_heap_terms_in_range(check_eterm,
+                                               bp->mem,
+                                               bp->mem + bp->used_size)) {
+                return 0;
+            }
+	    bp = bp->next;
+	}
+    }
+
+    return 1;
+}
+
+static int check_no_empty_boxed_non_literal_term(Eterm term) {
+    if (primary_tag(term) == TAG_PRIMARY_BOXED) {
+        Eterm empty_tuple_literal =
+            erts_get_global_literal(ERTS_LIT_EMPTY_TUPLE);
+        Eterm* term_boxed_ptr = boxed_val(term);
+        Eterm* literal_term_boxed_ptr = boxed_val(empty_tuple_literal);
+        Eterm arity_term = *term_boxed_ptr;
+        Uint arity = header_arity(arity_term);
+
+        if (arity == 0 &&
+            /*
+              MAP_SUBTAG terms can have 0 arity even though they have
+              something after the arity word.
+             */
+            (((arity_term) & _HEADER_SUBTAG_MASK) != MAP_SUBTAG) &&
+            term_boxed_ptr != literal_term_boxed_ptr) {
+            /*
+              Empty tuples are the only type of boxed value that can
+              have an arity of 0. This can change in the feature and
+              the condition above needs to be changed if it does.
+             */
+            erts_exit(ERTS_ABORT_EXIT,
+                      "Non-literal empty tuple found in heap.\n"
+                      "This is not allowed due to an optimization\n"
+                      "that assumes that the word after the arity\n"
+                      "word is allocated.\n");
+        }
+    }
```

Comment:
I found this hard to follow, how about something like:
```suggestion
   if (is_boxed(term)) {
        Uint arity = arityval(*boxed_val(term));

        /* Maps can have 0 arity even though they have something after the
         * arity word. */
        if (arity == 0 && !is_map(term)) {
            if (term != erts_get_global_literal(ERTS_LIT_EMPTY_TUPLE)) {
                /* Empty tuples are the only type of boxed value that can
                 * have an arity of 0. This can change in the feature and
                 * the condition above needs to be changed if it does. */
                erts_exit(ERTS_ABORT_EXIT,
                          "Non-literal empty tuple found in heap.\n"
                          "This is not allowed due to an optimization\n"
                          "that assumes that the word after the arity\n"
                          "word is allocated.\n");
            }
        }
    }
```

Reply:
Thanks, your code is easier to follow in my opinion as well.